### ISSUE TITLE
Remove-Item removes the symbolic links and junctions when used with -WhatIf parameter

### ISSUE THREAD

**Issue by Tospaa** [PowerShell/PowerShell#16664]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

- Create a new junction (reparse point) of an arbitrary directory
```powershell
New-Item -Path junction -ItemType Junction -Value 'C:\Program Files\'
```
- Create a new symbolic link of an arbitrary directory
```powershell
New-Item -Path symlink -ItemType SymbolicLink -Value 'C:\Program Files\'
```
- Run the `Remove-Item` command with the `-WhatIf` switch parameter
```powershell
Remove-Item -Path .\junction\ -WhatIf
Remove-Item -Path .\symlink\ -WhatIf
```

### Expected behavior

```console
Should not remove junction or symbolic link
```

### Actual behavior

```console
Removes the junction and symbolic link
```

### Environment data

```powershell
PS C:\Users\tospaa> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.2.1
PSEdition                      Core
GitCommitId                    7.2.1
OS                             Microsoft Windows 10.0.19043
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

https://user-images.githubusercontent.com/24327489/147056764-64cec441-7bf8-4657-bac7-7ca810df7117.mp4

**Comment by darthwalsh** [comment]
#19714 seems to be the same issue. (probably best to close one as a dupe?)

**Comment by msftrncs** [comment]
@jhoneill, the conversation appears to have moved to this older thread that didn't turn up when originally researching my issue.

**Comment by ViperTG** [comment]
Can confirm this is still an issue in 7.3.6, just had it delete my links while testing .. very annoying.

**Comment by msftrncs** [comment]
@ViperTG, still an issue, until this issue is closed.  @jhoneill had at least marked #19714 as a bug.

**Comment by jhoneill** [comment]
@msftrncs  I now longer have any influence to get items looked at having been removed from the working group.

**Comment by jhoneill** [comment]
> > @msftrncs I now longer have any influence to get items looked at having been removed from the working group.
> 
> Sorry to hear that. Interpersonal issues?

I got suspended on here for arguing with someone and categorising their behaviour over a long period as that of a troll. My access to the working group - and that of another member who'd be on the same side of the argument - vanished later without a word.  Characterize that as you will.

**Comment by jhoneill** [comment]
@ImportTaste  There seems to a small flurry of issues around the behaviour of `Remove-Item`, not always honouring `-whatif` is one, `-filter` and `-recurse` interacting in a way that is probably just wrong, but definitely counter-intuitive, and not having the same dynamic parameters as `Get-ChildItem`   (like `-file` and -`directory`) 

Someone more optimistic than I am might look to all these to be bundled up and addressed together.

**Comment by MatejKafka** [comment]
The issue seems specific to directory reparse points. For file symlinks, it works as expected.
```powershell
$null = mkdir dir
$null = ni file.txt

$null = ni -Type SymbolicLink filelink.txt -Target .\file.txt
$null = ni -Type SymbolicLink dirlink -Target .\dir

ls
rm .\filelink.txt -WhatIf
rm .\dirlink -WhatIf
ls
```

**Comment by MatejKafka** [comment]
The issue is here: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L3029-L3046

I don't have time to champion a PR, @JustinGrote said he might go for it.

Seems it was always somewhat broken, but this issue was exposed in https://github.com/PowerShell/PowerShell/commit/f1769fe7a80bcfc00b52e56e93a7ab8ed8b68c1c

**Comment by VirgilMing** [comment]
Disclaimer: This comment is mostly written by Codex/ChatGPT. I'm not exactly sure of the code review part, but I really observed the real world behavior.

Confirmed that this is still reproducible in PowerShell 7.6.4.

### Environment

```text
PSVersion       7.6.4
PSEdition       Core
GitCommitId     7.6.4
OS              Windows 11 Pro for Workstations, build 26200, x64
```

### Observed real-world behavior

During debugging with tooling that manages an NTFS junction, I deliberately used `-WhatIf` to preview the operation, with no intent of modifying the link:

```powershell
Test-Path -LiteralPath $junction
# output: True

Remove-Item -LiteralPath $junction -Recurse -Force -WhatIf -ErrorAction Stop

Test-Path -LiteralPath $junction
# output: False

Test-Path -LiteralPath $target
# output: True
```

PowerShell specifically printed:

```text
What if: Performing the operation "Remove Directory" on target "<junction>".
```

The invocation completed without error, but the junction itself was detached despite `-WhatIf`. The target directory and its contents remained intact.

This is particularly dangerous in recovery and diagnostic tooling, where `-WhatIf` is relied upon as the dry-run safety boundary before a destructive filesystem operation.

### Source-level cause in v7.6.4

The source-level cause previously identified in this thread remains present in v7.6.4:

https://github.com/PowerShell/PowerShell/blob/v7.6.4/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L3035-L3080

`RemoveDirectoryInfoItem()` stores the result of `ShouldProcess()` in `continueRemoval`:

```csharp
continueRemoval = ShouldProcess(directory.FullName, action);
```

However, the reparse-point/name-surrogate branch executes before the later `if (continueRemoval)` guard and calls:

```csharp
directory.Delete();
```

It then returns. Consequently, when `ShouldProcess()` returns `false` for `-WhatIf`, a junction or directory symbolic link is still detached.

The reparse-point branch should also be gated by `continueRemoval`, or the method should return immediately when `continueRemoval` is false.

A regression test should verify that:

* directory symbolic links remain present after `Remove-Item -WhatIf` on supported platforms;
* NTFS junctions remain present on Windows;
* the behavior is also correct with the real-world `-Recurse -Force -WhatIf` combination.

Since #19714 was labeled `Issue-Bug` and closed as a duplicate of this issue, could the same bug label also be applied to this primary tracking issue?
